### PR TITLE
Add checks against vault's 24h protection limits

### PR DESCRIPTION
**Pull request by StefanEx10**
This PR adds the necessary checks to prevent the operator from being able to take out all vault assets.

---

> [!NOTE]
> **Medium Risk**
> Adds new runtime checks that can revert `trade`, `liquidate`, `deleverage`, and vault redemption flows when a vault’s daily loss limit would be exceeded, which directly impacts core execution paths and could block operations if misconfigured.
> 
> **Overview**
> Introduces a **vault 24h protection-limit enforcement** by adding `Positions.validate_against_vault_limits` and wiring it into core state-changing flows (trades, liquidations, deleverages, and `redeem_from_vault`) using per-position configs from the `Vaults` component.
> 
> Hardens operator flows by **preventing withdrawals and transfers from vault positions** (`VAULT_CANNOT_WITHDRAW`/`VAULT_CANNOT_TRANSFER`), and updates flow tests to raise vault limits during redemption scenarios; minor spec formatting cleanup.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a4b66a3fb9471894d854465822625c9e43f14805. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237)

### CODE REVIEW

File: workspace/apps/perpetuals/contracts/src/core/components/vaults/vaults.cairo [hunk 1]
```diff
@@ -23,52 +29,137 @@ pub mod Vaults {
...
+            get_dep_component!(@self, Roles).only_app_governor();
+
+            let current_config = self.get_vault_config_for_position(:vault_position);
+            let positions = get_dep_component!(@self, Positions);
+            let position_tv_tr = positions.get_position_tv_tr(vault_position);
+            let tv_at_check = position_tv_tr.total_value;
+            let max_tv_loss = mul_wide_and_floor_div(
+                tv_at_check.abs(), percentage_basis_points.into() * 10, 1000,
+            )
+                .unwrap();
```
[the remark this answers is not in the document]
Comment (StefanIliev545) [reply workspace/apps/perpetuals/contracts/src/core/components/vaults/vaults.cairo]: Rebased on latest of the branch that adds the config so this should be resolved.

File: workspace/apps/perpetuals/contracts/src/core/components/positions/positions.cairo [hunk 2]
```diff
@@ -808,6 +809,33 @@ pub mod Positions {
...
+                                "Vault Protection Limit Exceeded, tv_at_last_check: {}, tv_after_operation: {}, max_allowed_loss : {}",
+                                tv_at_last_check,
+                                tv_after,
+                                max_tvtr_loss,
+                            ),
+                        );
+                    }
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (MohammadNassar1) [reply workspace/apps/perpetuals/contracts/src/core/components/positions/positions.cairo:777]: Can you add position_id to the error msg?

File: workspace/apps/perpetuals/contracts/src/core/components/transfer/transfer_manager.cairo [hunk 3]
```diff
@@ -151,6 +156,7 @@ pub(crate) mod TransferManager {
 
     component!(path: FulfillmentComponent, storage: fulfillment_tracking, event: FulfillmentEvent);
     component!(path: PausableComponent, storage: pausable, event: PausableEvent);
+    component!(path: SystemTimeComponent, storage: system_time, event: SystemTimeEvent);
```
[the remark this answers is not in the document]
Comment (MohammadNassar1) [reply workspace/apps/perpetuals/contracts/src/core/components/transfer/transfer_manager.cairo]: It's not an issue, vault position cannot transfer.

**Review by MohammadNassar1 — CHANGES_REQUESTED** [review]
> @MohammadNassar1 reviewed 10 files and all commit messages, and made 4 comments.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237)* status: all files reviewed, 4 unresolved discussions (waiting on StefanEx10 and StefanIliev545).

___
*[`workspace/apps/perpetuals/contracts/src/core/components/deleverage/deleverage_manager.cairo` line 404 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237#-Om4HxYx3loWyKB3Zl_2:-Om4HxYx3loWyKB3Zl_3:bcgct1q) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/c66d43afa884e474d84a1dadde9fd83ff56f22f1/workspace/apps/perpetuals/contracts/src/core/components/deleverage/deleverage_manager.cairo#L404)):*
> ```Cairo
>             // component at the very least to get the asset type but could be more depending on the
>             // validation logic).
> ```

Can you please add a comment explaining why you do not validate deleveraged_position?
___
*[`workspace/apps/perpetuals/contracts/src/core/components/withdrawal/withdrawal_manager.cairo` line 472 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237#-Om4KddNCSApBTCqPgjr:-Om4KddNCSApBTCqPgjs:b-dhn4v5) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/c66d43afa884e474d84a1dadde9fd83ff56f22f1/workspace/apps/perpetuals/contracts/src/core/components/withdrawal/withdrawal_manager.cairo#L472)):*
> ```Cairo
>             interest_amount: i64,
>         ) -> (HashType, ContractAddress) {
>             assert!(!self.vaults.is_vault_position(position_id), "VAULT_CANNOT_WITHDRAW");
> ```

It's already checked in withdraw request.



_Code quote:_
```Cairo
assert!(!self.vaults.is_vault_position(position_id), "VAULT_CANNOT_WITHDRAW");
```

**Comment by StefanEx10** [comment]
> It's already checked in withdraw request.
> 
> _Code quote:_
> 
> ```cairo
> assert!(!self.vaults.is_vault_position(position_id), "VAULT_CANNOT_WITHDRAW");
> ```

Someone can place a withdrawal request before a position is a vault and in the meanwhile it becomes.

**Comment by MohammadNassar1** [comment]
*[`workspace/apps/perpetuals/contracts/src/core/components/transfer/transfer_manager.cairo` line 335 at r5](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237#-Om9sXQo4_2GaO8j2hk2:-Om9sXQo4_2GaO8j2hk3:b11em9i) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/7aae7b2a3774ea508994e2432397d34967a9a873/workspace/apps/perpetuals/contracts/src/core/components/transfer/transfer_manager.cairo#L335)):*
> ```Cairo
>         ) {
>             // Parameters
>             assert!(!self.vaults.is_vault_position(position_id), "VAULT_CANNOT_TRANSFER");
> ```

Redundant.
It's checked in transfer_request.



_Code quote:_
```Cairo
assert!(!self.vaults.is_vault_position(position_id), "VAULT_CANNOT_TRANSFER");
```

**Comment by MohammadNassar1** [comment]
*[`workspace/apps/perpetuals/contracts/src/core/components/vaults/events.cairo` line 152 at r5](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237#-Om9soydC1vE7KCSB0iS:-Om9soydC1vE7KCSB0iT:brexe62) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/7aae7b2a3774ea508994e2432397d34967a9a873/workspace/apps/perpetuals/contracts/src/core/components/vaults/events.cairo#L152)):*
> ```Cairo
>     pub new_tv_at_check: i128,
>     pub new_max_tv_loss: u128,
> }
> ```

Please remove, this struct appears twice.



_Code quote:_
```Cairo
#[derive(Debug, Drop, PartialEq, starknet::Event)]
pub struct VaultProtectionReset {
    #[key]
    pub vault_position_id: PositionId,
    pub old_tv_at_check: i128,
    pub old_max_tv_loss: u128,
    pub new_tv_at_check: i128,
    pub new_max_tv_loss: u128,
}
```

**Comment by StefanEx10** [comment]
> _[`workspace/apps/perpetuals/contracts/src/core/components/transfer/transfer_manager.cairo` line 335 at r5](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237#-Om9sXQo4_2GaO8j2hk2:-Om9sXQo4_2GaO8j2hk3:b11em9i) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/7aae7b2a3774ea508994e2432397d34967a9a873/workspace/apps/perpetuals/contracts/src/core/components/transfer/transfer_manager.cairo#L335)):_
> 
> > ```cairo
> >         ) {
> >             // Parameters
> >             assert!(!self.vaults.is_vault_position(position_id), "VAULT_CANNOT_TRANSFER");
> > ```
> 
> Redundant. It's checked in transfer_request.
> 
> _Code quote:_
> 
> ```cairo
> assert!(!self.vaults.is_vault_position(position_id), "VAULT_CANNOT_TRANSFER");
> ```

The logic here is the same as with withdrawals - a request can be placed before a position becomes marked as a vault position, but the execution can come after. Edge case race condition.

**Review by MohammadNassar1 — CHANGES_REQUESTED** [review]
> @MohammadNassar1 partially reviewed 2 files and made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237)* status: 1 of 11 files reviewed, 6 unresolved discussions (waiting on StefanEx10 and StefanIliev545).

___
*[`workspace/apps/perpetuals/contracts/src/core/components/deleverage/deleverage_manager.cairo` line 427 at r6](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237#-OmFFEiX-sFcjqkOVSdW:-OmFFEiX-sFcjqkOVSdX:b-8txmq2) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/afbbd05c60782cde24bb118508e5b643b37906d8/workspace/apps/perpetuals/contracts/src/core/components/deleverage/deleverage_manager.cairo#L427)):*
> ```Cairo
>                         .get_vault_protection_config(deleverager_position_id),
>                     :tvtr,
>                 );
> ```

Where do you update the new values of vault config (tv, and loss)



_Code quote:_
```Cairo
            self
                .positions
                .validate_against_vault_limits(
                    position_id: deleverager_position_id,
                    vault_protection_config: self
                        .vaults
                        .get_vault_protection_config(deleverager_position_id),
                    :tvtr,
                );
```

**Comment by StefanEx10** [comment]
> > @MohammadNassar1 partially reviewed 2 files and made 1 comment.
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237)_ status: 1 of 11 files reviewed, 6 unresolved discussions (waiting on StefanEx10 and StefanIliev545).
> 
> _[`workspace/apps/perpetuals/contracts/src/core/components/deleverage/deleverage_manager.cairo` line 427 at r6](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237#-OmFFEiX-sFcjqkOVSdW:-OmFFEiX-sFcjqkOVSdX:b-8txmq2) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/afbbd05c60782cde24bb118508e5b643b37906d8/workspace/apps/perpetuals/contracts/src/core/components/deleverage/deleverage_manager.cairo#L427)):_
> 
> > ```cairo
> >                         .get_vault_protection_config(deleverager_position_id),
> >                     :tvtr,
> >                 );
> > ```
> 
> Where do you update the new values of vault config (tv, and loss)
> 
> _Code quote:_
> 
> ```cairo
>             self
>                 .positions
>                 .validate_against_vault_limits(
>                     position_id: deleverager_position_id,
>                     vault_protection_config: self
>                         .vaults
>                         .get_vault_protection_config(deleverager_position_id),
>                     :tvtr,
>                 );
> ```


the limit is updated in the get_vault_protection_config once per day;
The TV is tracked there for the beginning of the day on the first call and subsequent withdraws validate that they dont leave the TV below an allowed limit compared to start of the day.

File: workspace/apps/perpetuals/contracts/src/core/components/liquidation/liquidation_manager.cairo [hunk 4]
```diff
@@ -606,14 +613,23 @@ pub(crate) mod LiquidationManager {
...
                 );
+            self
+                .positions
+                .validate_against_vault_limits(
+                    position_id: liquidator_position_id,
+                    vault_protection_config: self
+                        .vaults
+                        .get_vault_protection_config(liquidator_position_id),
+                    :tvtr,
+                );
```
[the remark this answers is not in the document]
Comment (MohammadNassar1) [reply workspace/apps/perpetuals/contracts/src/core/components/liquidation/liquidation_manager.cairo:632]: It's not an issue, the receiver can only be healthier.

**Review by MohammadNassar1 — COMMENTED** [review]
> @MohammadNassar1 reviewed 10 files and all commit messages, made 1 comment, and resolved 4 discussions.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237)* status: all files reviewed, 2 unresolved discussions (waiting on StefanEx10).

**Review by MohammadNassar1 — APPROVED** [review]
> @MohammadNassar1 resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237)* status: all files reviewed, 1 unresolved discussion (waiting on StefanEx10).

**Review by MohammadNassar1 — APPROVED** [review]
> @MohammadNassar1 reviewed 7 files and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on StefanEx10).

**Review by MohammadNassar1 — COMMENTED** [review]
> @MohammadNassar1 reviewed 2 files.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237)* status: all files reviewed (commit messages unreviewed), all discussions resolved (waiting on StefanEx10).

**Review by MohammadNassar1 — APPROVED** [review]
> @MohammadNassar1 reviewed 9 files and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on StefanEx10).

**Review by MohammadNassar1 — CHANGES_REQUESTED** [review]
> @MohammadNassar1 made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237)* status: all files reviewed, 1 unresolved discussion (waiting on StefanEx10).

___
*[`workspace/apps/perpetuals/contracts/src/core/components/vaults/vaults_contract.cairo` line 733 at r11](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237#-OmLES9X0v-Jr_hd9IlV:-OmLES9YEhPLC6YHrCBR:b-plwr77) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/33f8cf643e45468e631dc6ffc5cf7fd74fc9c8ef/workspace/apps/perpetuals/contracts/src/core/components/vaults/vaults_contract.cairo#L733)):*
> ```Cairo
>                         .get_vault_protection_config(vault_position_id),
>                     :tvtr,
>                 );
> ```

Please add to `invest_in_vault` flow as well.



_Code quote:_
```Cairo
            self
                .positions
                .validate_against_vault_limits(
                    position_id: vault_position_id,
                    vault_protection_config: self
                        .vaults
                        .get_vault_protection_config(vault_position_id),
                    :tvtr,
                );
```

**Review by MohammadNassar1 — COMMENTED** [review]
> @MohammadNassar1 reviewed 3 files and all commit messages, and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237)* status: 14 of 15 files reviewed, all discussions resolved (waiting on StefanIliev545).

**Review by MohammadNassar1 — APPROVED** [review]
> @MohammadNassar1 reviewed 1 file.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/237)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on StefanEx10).
